File: new_backend/express-backend/src/routes/brands.ts
import express, { Request, Response, Router } from 'express';
import Joi from 'joi';
import { PrismaClient } from '@prisma/client';

const router: Router = express.Router();

// ── Validation ────────────────────────────────────────────────────────────────

const policyRuleSchema = Joi.object({
  type: Joi.string().valid('allow', 'restrict', 'require').required(),
  target: Joi.string().valid('topic', 'documentSource', 'documentTag').required(),
  value: Joi.string().required(),
});

const createBrandSchema = Joi.object({
  name: Joi.string().min(1).max(200).required(),
  brandVoice: Joi.string().max(2000).allow('', null).optional(),
  allowedTopics: Joi.array().items(Joi.string()).optional(),
  restrictedTopics: Joi.array().items(Joi.string()).optional(),
  policyRules: Joi.array().items(policyRuleSchema).optional(),
});

const updateBrandSchema = Joi.object({
  name: Joi.string().min(1).max(200).optional(),
  brandVoice: Joi.string().max(2000).allow('', null).optional(),
  allowedTopics: Joi.array().items(Joi.string()).optional(),
  restrictedTopics: Joi.array().items(Joi.string()).optional(),
  policyRules: Joi.array().items(policyRuleSchema).optional(),
});

// ── GET /api/brands ───────────────────────────────────────────────────────────
// List all brands for the authenticated tenant.
router.get('/', async (req: Request, res: Response) => {
  try {
    const prisma: PrismaClient = req.app.get('prisma');
    const brands = await prisma.brand.findMany({
      where: { tenantId: req.tenantId },
      include: {
        _count: { select: { agents: true } },
      },
      orderBy: { createdAt: 'desc' },
    });
    res.json(brands);
  } catch (error) {
    console.error('Error fetching brands:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// ── GET /api/brands/:id ───────────────────────────────────────────────────────
router.get('/:id', async (req: Request, res: Response) => {
  try {
    const prisma: PrismaClient = req.app.get('prisma');
    const brand = await prisma.brand.findFirst({
      where: { id: req.params.id, tenantId: req.tenantId },
      include: {
        agents: { select: { id: true, name: true, status: true } },
      },
    });
    if (!brand) return res.status(404).json({ error: 'Brand not found' });
    res.json(brand);
  } catch (error) {
    console.error('Error fetching brand:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// ── POST /api/brands ──────────────────────────────────────────────────────────
router.post('/', async (req: Request, res: Response) => {
  const { error, value } = createBrandSchema.validate(req.body);
  if (error) return res.status(400).json({ error: error.details[0].message });

  try {
    const prisma: PrismaClient = req.app.get('prisma');
    const brand = await prisma.brand.create({
      data: {
        tenantId: req.tenantId,
        name: value.name,
        brandVoice: value.brandVoice ?? null,
        allowedTopics: value.allowedTopics ?? [],
        restrictedTopics: value.restrictedTopics ?? [],
        policyRules: value.policyRules ?? [],
      },
    });
    res.status(201).json(brand);
  } catch (error) {
    console.error('Error creating brand:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// ── PUT /api/brands/:id ───────────────────────────────────────────────────────
router.put('/:id', async (req: Request, res: Response) => {
  const { error, value } = updateBrandSchema.validate(req.body);
  if (error) return res.status(400).json({ error: error.details[0].message });

  try {
    const prisma: PrismaClient = req.app.get('prisma');

    const existing = await prisma.brand.findFirst({
      where: { id: req.params.id, tenantId: req.tenantId },
    });
    if (!existing) return res.status(404).json({ error: 'Brand not found' });

    const brand = await prisma.brand.update({
      where: { id: req.params.id },
      data: value,
    });
    res.json(brand);
  } catch (error) {
    console.error('Error updating brand:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// ── DELETE /api/brands/:id ────────────────────────────────────────────────────
router.delete('/:id', async (req: Request, res: Response) => {
  try {
    const prisma: PrismaClient = req.app.get('prisma');

    const existing = await prisma.brand.findFirst({
      where: { id: req.params.id, tenantId: req.tenantId },
    });
    if (!existing) return res.status(404).json({ error: 'Brand not found' });

    await prisma.brand.delete({ where: { id: req.params.id } });
    res.json({ deleted: true });
  } catch (error) {
    console.error('Error deleting brand:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

export default router;

File: python/backend/app/routes/brands.py
"""
/api/brands routes — mirrors Express src/routes/brands.ts
"""
from fastapi import APIRouter, Depends
from fastapi.responses import JSONResponse
from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.auth import AuthContext, get_auth
from app.models import Brand, Agent

router = APIRouter()


def _brand_dict(b: Brand, agent_count: int = 0) -> dict:
    return {
        "id": b.id,
        "tenantId": b.tenantId,
        "name": b.name,
        "brandVoice": b.brandVoice,
        "allowedTopics": b.allowedTopics,
        "restrictedTopics": b.restrictedTopics,
        "policyRules": b.policyRules,
        "createdAt": b.createdAt.isoformat() if b.createdAt else None,
        "updatedAt": b.updatedAt.isoformat() if b.updatedAt else None,
        "_count": {"agents": agent_count},
    }


@router.get("/")
async def list_brands(auth: AuthContext = Depends(get_auth), db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Brand).where(Brand.tenantId == auth.tenant_id).order_by(Brand.createdAt.desc()))
    brands = result.scalars().all()

    brand_ids = [b.id for b in brands]
    counts: dict[str, int] = {}
    if brand_ids:
        cr = await db.execute(
            select(Agent.brandId, func.count(Agent.id)).where(Agent.brandId.in_(brand_ids)).group_by(Agent.brandId)
        )
        counts = dict(cr.all())

    return [_brand_dict(b, counts.get(b.id, 0)) for b in brands]


@router.get("/{brand_id}")
async def get_brand(brand_id: str, auth: AuthContext = Depends(get_auth), db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Brand).where(Brand.id == brand_id, Brand.tenantId == auth.tenant_id))
    brand = result.scalar_one_or_none()
    if not brand:
        return JSONResponse({"error": "Brand not found"}, status_code=404)

    agents_r = await db.execute(
        select(Agent.id, Agent.name, Agent.status).where(Agent.brandId == brand_id)
    )
    agents = [{"id": r[0], "name": r[1], "status": r[2]} for r in agents_r.all()]

    d = _brand_dict(brand)
    d["agents"] = agents
    return d


@router.post("/")
async def create_brand(body: dict, auth: AuthContext = Depends(get_auth), db: AsyncSession = Depends(get_db)):
    name = body.get("name")
    if not name:
        return JSONResponse({"error": "\"name\" is required"}, status_code=400)

    brand = Brand(
        tenantId=auth.tenant_id,
        name=name,
        brandVoice=body.get("brandVoice"),
        allowedTopics=body.get("allowedTopics", []),
        restrictedTopics=body.get("restrictedTopics", []),
        policyRules=body.get("policyRules", []),
    )
    db.add(brand)
    await db.flush()
    await db.commit()
    await db.refresh(brand)
    return JSONResponse(_brand_dict(brand), status_code=201)


@router.put("/{brand_id}")
async def update_brand(brand_id: str, body: dict, auth: AuthContext = Depends(get_auth), db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Brand).where(Brand.id == brand_id, Brand.tenantId == auth.tenant_id))
    brand = result.scalar_one_or_none()
    if not brand:
        return JSONResponse({"error": "Brand not found"}, status_code=404)

    for field in ("name", "brandVoice", "allowedTopics", "restrictedTopics", "policyRules"):
        if field in body:
            setattr(brand, field, body[field])
    await db.commit()
    await db.refresh(brand)
    return _brand_dict(brand)


@router.delete("/{brand_id}")
async def delete_brand(brand_id: str, auth: AuthContext = Depends(get_auth), db: AsyncSession = Depends(get_db)):
    result = await db.execute(select(Brand).where(Brand.id == brand_id, Brand.tenantId == auth.tenant_id))
    brand = result.scalar_one_or_none()
    if not brand:
        return JSONResponse({"error": "Brand not found"}, status_code=404)
    await db.delete(brand)
    await db.commit()
    return {"deleted": True}
